### DISCUSSION TITLE
SmoothControls: avoid checks on event.type and touchMode in onPointerMove

### DISCUSSION THREAD

**Discussion by lucadalli** [google/model-viewer#2687] [General]
What is the sentiment towards moving away from the checks on `event.type` and `touchMode` in `onPointerMove` and instead attaching the correct switch subroutine directly as a `mousemove`/`touchmove` handler depending on `pointerdown` event type (`mousedown` or `touchdown`) and the number of touches?

This would roughly equate to the `mousedown` handler attaching the `mousemove` handler with `handleSinglePointerMove` and a `mouseup` handler which detaches the `mousemove` handler and the `mouseup` handler itself.

For touch events, the `pointermove` event handler is determined every time a pointer is added or removed. So instead of `touchMode`, the manipulated variable is a new one named `touchmoveHandler` which is assigned the currently active `touchmove` event listener so that it can be detached when `!touches.length` or if the element is removed from the DOM mid-interaction.

An added benefit of this approach is that the `mouseup` and `touchend` listeners are not added prematurely. Currently, these handlers are added as soon as controls are enabled, meaning that they sometimes fire for events which are unrelated to a model-viewer.
With the proposed approach the only two persistent handlers would be `touchstart` and `mousedown` on the model-viewer element. The rest are attached and detached in those two on the fly.

Is the current implementation done this way to facilitate a migration to pointer events down the line?
Considering that the code is peppered with checks about the pointer type, would it be favorable to migrate to pointer events?

**Comment by elalish** [comment]
That all sounds fairly reasonable to me, though obviously we'll need to do some code review and lots of testing. Regarding pointer events, you should look over an old attempt I made: https://github.com/google/model-viewer/pull/1645. I was blocked by non-standard iOS Safari behavior I couldn't find a work around for. I suppose it's possible your other changes here might help avoid that bug, but it's hard to say. Especially if you don't have an iOS device for local testing, I'd stay away from pointer events for now.

**Reply by lucadalli** [reply]
Cool, I'd be keen to work on this but I'm facing some difficulties getting the tooling to work on Win 10. I'm using WSL2 as advised but the tests fail to run and the `dev` command does not seem to build. I'm having to rebuild manually after every change :(

I've tried gone through #1121 and its linked issues, to no avail. I'll continue battling with it tomorrow. Fingers crossed. I'm going to open a discussion with some questions I have.

**Reply by elalish** [reply]
Yes, for sure. Also we can loop in @timmmeh as I believe he most recently got a Windows dev environment set up.

**Reply by lucadalli** [reply]
Since I my handicap is not as severe as I thought and I managed to run the tests (#2690), I'd like to get started on this task, followed by #2684 and #2682 . That order should minimize re-work.

**Reply by elalish** [reply]
Sounds like a good plan to me. Thanks for the help!
